mods/d2k/maps/shellmap/d2k-shellmap.lua

```diff
@@ -0,0 +1,188 @@
+EmperorStarportTypes = { "trike.starport", "quad.starport", "siege_tank.starport", "missile_tank.starport", "combat_tank_h.starport" }
+
+Upgrades = { "upgrade.barracks", "upgrade.light", "upgrade.conyard", "upgrade.heavy", "upgrade.hightech" }
+
+Harvester = { "harvester" }
+
+AtrCarryHarvWaypoints = { atr_harvcarry_2.Location, atr_harvcarry_1.Location }
+HarCarryHarvWaypoints = { har_harvcarry_2.Location, har_harvcarry_1.Location }
+OrdCarryHarvWaypoints = { ord_harvcarry_2.Location, ord_harvcarry_1.Location }
+EmpCarryHarvWaypoints = { emp_harvcarry_2.Location, emp_harvcarry_1.Location }
+SmgCarryHarvWaypoints = { smg_harvcarry_2.Location, smg_harvcarry_1.Location }
+
+IdleHunt = function(unit) if not unit.IsDead then Trigger.OnIdle(unit, unit.Hunt) end end
+
+SendAttack = function(house)
+	if IsAttacking[house] then
+		return
+	end
+	IsAttacking[house] = true
+	HoldProduction[house] = true
+
+	local units = SetupAttackGroup(house)
+	Utils.Do(units, function(unit)
+		IdleHunt(unit)
+	end)
+
+	Trigger.OnAllRemovedFromWorld(units, function()
+		IsAttacking[house] = false
+		HoldProduction[house]  = false
+	end)
+end
+
+SetupAttackGroup = function(house)
+	local units = { }
+
+	for i = 0, AttackGroupSize[1], 1 do
+		if #IdlingUnits[house] == 0 then
+			return units
+		end
+
+		local number = Utils.RandomInteger(1, #IdlingUnits[house])
+
+		if IdlingUnits[house][number] and not IdlingUnits[house][number].IsDead then
+			units[i] = IdlingUnits[house][number]
+			table.remove(IdlingUnits[house], number)
+		end
+	end
+
+	return units
+end
+
+SendNewHarv = function(house, waypoint, count)
+	local harvesters = house.GetActorsByType("harvester")
+	if #harvesters < count then
+		Reinforcements.ReinforceWithTransport(house, "carryall.reinforce", Harvester , waypoint, { waypoint[1] })
+		Trigger.AfterDelay(DateTime.Seconds(45), function()
+			SendNewHarv(house, waypoint, count)
+		end)
+	end
+end
+
+ticks = 0
+speed = 5
+
+Tick = function()
+	ticks = ticks + 1
+
+	local t = (ticks + 45) % (360 * speed) * (math.pi / 180) / speed;
+	Camera.Position = viewportOrigin + WVec.New(19200 * math.sin(t), 28800 * math.cos(t), 0)
+end
+
+WorldLoaded = function()
+	atreides = Player.GetPlayer("Atreides")
+	harkonnen = Player.GetPlayer("Harkonnen")
+	ordos = Player.GetPlayer("Ordos")
+	emperor = Player.GetPlayer("Emperor")
+	smugglers = Player.GetPlayer("Smugglers")
+	
+	viewportOrigin = Camera.Position
+	
+	atr_cyard.Produce(Upgrades[1])
+	atr_cyard.Produce(Upgrades[2])
+	atr_cyard.Produce(Upgrades[3])
+	atr_cyard.Produce(Upgrades[4])
+	atr_cyard.Produce(Upgrades[5])
+	har_cyard.Produce(Upgrades[1])
+	har_cyard.Produce(Upgrades[2])
+	har_cyard.Produce(Upgrades[3])
+	har_cyard.Produce(Upgrades[4])
+	ord_cyard.Produce(Upgrades[1])
+	ord_cyard.Produce(Upgrades[2])
+	ord_cyard.Produce(Upgrades[3])
+	ord_cyard.Produce(Upgrades[4])
+	emp_cyard.Produce(Upgrades[1])
+	emp_cyard.Produce(Upgrades[2])
+	emp_cyard.Produce(Upgrades[3])
+	emp_cyard.Produce(Upgrades[4])
```

Comment:
This can be simplified to something like:
```
Utils.Do(Utils.Take(4, Upgrades), function(upgrade)
	atr_cyard.Produce(upgrade)
	har_cyard.Produce(upgrade)
	ord_cyard.Produce(upgrade)
	emp_cyard.Produce(upgrade)
end)
atr_cyard.Produce(Upgrades[5])